// firebase-tools/lib/emulator/events/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("lodash");
class EventUtils {
    static isEvent(proto) {
        return _.has(proto, "context") && _.has(proto, "data");
    }
    static isLegacyEvent(proto) {
        return _.has(proto, "data") && _.has(proto, "resource");
    }
    static convertFromLegacy(event, service) {
        return {
            context: {
                eventId: event.eventId || "",
                timestamp: event.timestamp || "",
                eventType: event.eventType || "",
                resource: {
                    name: event.resource || "",
                    service,
                },
            },
            data: event.data,
        };
    }
    static convertToLegacy(event) {
        return {
            eventId: event.context.eventId,
            timestamp: event.context.timestamp,
            eventType: event.context.eventType,
            resource: event.context.resource.name,
            data: event.data,
        };
    }
}
exports.EventUtils = EventUtils;
